tagformat/caseconversion/case_conversion.go

```diff
@@ -26,6 +27,231 @@ type EncodeCasingFunc func(DecodedIdentifier) string
+
+	lowerCased := make([]string, 0, len(words))
+
+	// see if any of the initialisms are actually a combination of two ("JSONAPI" or something...)
+	for _, word := range words {
+		if strings.ToUpper(word) != word {
+			// it's not an initialism because it's not all uppercase
+			lowerCased = append(lowerCased, strings.ToLower(word))
+			continue
+		}
+		chunkLen := 2
+		for chunkLen <= len(word) {
+			offset := sort.SearchStrings(g.initialisms, word[:chunkLen])
+			if offset < len(g.initialisms) && g.initialisms[offset] == word[:chunkLen] {
+				// found a match
+				lowerCased = append(lowerCased, strings.ToLower(word[:chunkLen]))
+				word = word[chunkLen:]
+				chunkLen = 2
+				continue
+			}
+			chunkLen++
+		}
+		// if there's anything left to the word, add it
+		if len(word) > 0 {
+			lowerCased = append(lowerCased, strings.ToLower(word))
+		}
+	}
+	return lowerCased, nil
+}
+
+// Encode implements a EncodeCasingFunc for Go-style identifiers and returns an
+// exported-style name (with an initial uppercase character).
+func (g *GoCaseConverter) Encode(words DecodedIdentifier) string {
+	b := strings.Builder{}
+	b.Grow(aggregateStringLen(words))
+	for _, w := range words {
+		maybeInitialism := strings.ToUpper(w)
+		initialismOffset := sort.SearchStrings(g.initialisms, maybeInitialism)
+
+		atomOffset := sort.SearchStrings(g.loweredAtoms, w)
+
+		// check first if it's an atom, then if it's an initialism, and then
+		// assume it's just a normal name.
+		if atomOffset < len(g.atoms) && g.loweredAtoms[atomOffset] == w {
+			b.WriteString(g.atoms[atomOffset])
+		} else if initialismOffset < len(g.initialisms) && g.initialisms[initialismOffset] == maybeInitialism {
+			b.WriteString(maybeInitialism)
+		} else {
+			b.WriteString(cases.Title(language.English, cases.NoLower).String(w))
+		}
+	}
+	return b.String()
+}
+
+// EncodeUnexported is like Encode, but returns an unexported name (with an
+// initial lowercase character) still adhering to the rules of initialisms and
+// atoms.
+func (g *GoCaseConverter) EncodeUnexported(words DecodedIdentifier) string {
+	b := strings.Builder{}
+	b.Grow(aggregateStringLen(words))
+	for i, w := range words {
+		maybeInitialism := strings.ToUpper(w)
+		initialismOffset := sort.SearchStrings(g.initialisms, maybeInitialism)
+
+		atomOffset := sort.SearchStrings(g.loweredAtoms, w)
+
+		// check first if it's an atom, then if it's an initialism, and then
+		// assume it's just a normal name.
+		if atomOffset < len(g.atoms) && g.loweredAtoms[atomOffset] == w {
+			if i == 0 {
+				b.WriteString(g.loweredAtoms[atomOffset])
+			} else {
+				b.WriteString(g.atoms[atomOffset])
+			}
+		} else if initialismOffset < len(g.initialisms) && g.initialisms[initialismOffset] == maybeInitialism {
+			if i == 0 {
+				b.WriteString(w)
+			} else {
+				b.WriteString(maybeInitialism)
+			}
+		} else {
+			if i == 0 {
+				b.WriteString(w)
+			} else {
+				b.WriteString(cases.Title(language.English, cases.NoLower).String(w))
```

Comment:
seems like you can pull the `i == 0` check here up into the outer `if`/`else` cascade and simplify both branches.

Reply:
While there is some duplication in statements I feel like the original way I wrote it is easier to understand -- but I've updated it nonetheless to simplify the branches.

Reply:
yeah, it might warrant a comment.
It is a little odd to see that `i == 0` check at the beginning  now.